### PR TITLE
Run3-sim48 First step to remove access of DDCompactView in SimG4CMS/Calo for ECAL

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Create objects for ECAL which are needed at SIM step from specpar etc so that SIM step does not require access to DDCompactView and can use this object. The change in SIM step to be done in the next step 

#### PR validation:

Tested with standard workflows

#### if this PR is a backport please specify the original PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
Using CMSSW_11_0_X_2019-11-03-2300
```
cmsRun Geometry/EcalCommonData/test/python/runEcalSimParameterDD4Hep_cfg.py
```
gives an error:
```
----- Begin Fatal Exception 05-Nov-2019 11:04:22 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'p1'
   [2] Calling method for module EcalSimParametersAnalyzer/'ecalSimulationParametersAnalyzerEB'
   [3] Using EventSetup component EcalSimParametersESModule/'ecalSimulationParametersEB' to make data EcalSimulationParameters/'EcalHitsEB' in record IdealGeometryRecord
   [4] Using EventSetup component DDCompactViewESProducer/'' to make data DDCompactView/'' in record IdealGeometryRecord
   [5] Using EventSetup component DDDetectorESProducer/'' to make data DDDetector/'' in record IdealGeometryRecord
Exception Message:
A std::exception was thrown.
An exception of category 'FatalRootError' occurred.
   Additional Info:
      [a] Fatal Root Error: @SUB=TGeoVoxelFinder::SortAll
Volume world_volume: Cannot make slices on any axis


dd4hep: while parsing /build/yana/DD4HEP/Sunanda/CMSSW_11_0_X_2019-11-03-2300/src/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml
dd4hep: with plugin:DD4hep_CompactLoader
----- End Fatal Exception -------------------------------------------------
```

**Comment by ianna** [comment]
> Using CMSSW_11_0_X_2019-11-03-2300
> 
> ```
> cmsRun Geometry/EcalCommonData/test/python/runEcalSimParameterDD4Hep_cfg.py
> ```
> 
> gives an error:
> 
> ```
> ----- Begin Fatal Exception 05-Nov-2019 11:04:22 CET-----------------------
> An exception of category 'StdException' occurred while
>    [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
>    [1] Running path 'p1'
>    [2] Calling method for module EcalSimParametersAnalyzer/'ecalSimulationParametersAnalyzerEB'
>    [3] Using EventSetup component EcalSimParametersESModule/'ecalSimulationParametersEB' to make data EcalSimulationParameters/'EcalHitsEB' in record IdealGeometryRecord
>    [4] Using EventSetup component DDCompactViewESProducer/'' to make data DDCompactView/'' in record IdealGeometryRecord
>    [5] Using EventSetup component DDDetectorESProducer/'' to make data DDDetector/'' in record IdealGeometryRecord
> Exception Message:
> A std::exception was thrown.
> An exception of category 'FatalRootError' occurred.
>    Additional Info:
>       [a] Fatal Root Error: @SUB=TGeoVoxelFinder::SortAll
> Volume world_volume: Cannot make slices on any axis
> 
> 
> dd4hep: while parsing /build/yana/DD4HEP/Sunanda/CMSSW_11_0_X_2019-11-03-2300/src/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml
> dd4hep: with plugin:DD4hep_CompactLoader
> ----- End Fatal Exception -------------------------------------------------
> ```

To fix this, please, update:
```
$ git diff
diff --git a/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml b/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml
index c86b041cd8b..f9abd77feb7 100644
--- a/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml
+++ b/Geometry/EcalCommonData/data/dd4hep/cms-ecal-geometry.xml
@@ -63,11 +63,4 @@
     <Include ref='Geometry/EcalCommonData/data/esalgo/2021/v1/esalgo.xml'/>
     <Include ref='Geometry/TrackerCommonData/data/trackermaterial.xml'/>
   </IncludeSection>
-  
-  <PosPartSection label="">
-    <PosPart copyNumber="1">
-      <rParent name=":world_volume"/>
-      <rChild name="cms:OCMS"/>
-    </PosPart>
-  </PosPartSection>
 </DDDefinition>

```

File: Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "DetectorDescription/Core/interface/DDCompactView.h"
+#include "DetectorDescription/DDCMS/interface/DDCompactView.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <string>
+#include <vector>
+
+class DDFilteredView;
+class EcalSimulationParameters;
+
+template <typename T>
```
Comment (ianna) [inline Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h]: @bsunanda , please do not expose it. One way to do it is to move it to cc:
```
diff --git a/Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h b/Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h
index 80623ac215a..cc7c80e0c02 100644
--- a/Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h
+++ b/Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h
@@ -11,16 +11,6 @@
 class DDFilteredView;
 class EcalSimulationParameters;
 
-template <typename T>
-void myPrint(std::string value, const std::vector<T>& vec) {
-  edm::LogVerbatim("HCalGeom") << "EcalSimParametersFromDD: " << vec.size() << " entries for " << value << ":";
-  unsigned int i(0);
-  for (const auto& e : vec) {
-    edm::LogVerbatim("HCalGeom") << " (" << i << ") " << e;
-    ++i;
-  }
-}
-
 class EcalSimParametersFromDD {
 public:
   EcalSimParametersFromDD() = default;
diff --git a/Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc b/Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc
index c4b8f3e6e83..379ef3d2eee 100644
--- a/Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc
+++ b/Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc
@@ -9,6 +9,17 @@
 #include <iomanip>
 
 //#define EDM_ML_DEBUG
+namespace {
+  template <typename T>
+  void myPrint(std::string value, const std::vector<T>& vec) {
+    edm::LogVerbatim("HCalGeom") << "EcalSimParametersFromDD: " << vec.size() << " entries for " << value << ":";
+    unsigned int i(0);
+    for (const auto& e : vec) {
+      edm::LogVerbatim("HCalGeom") << " (" << i << ") " << e;
+      ++i;
+    }
+  }
+}  // namespace
 
 bool EcalSimParametersFromDD::build(const DDCompactView* cpv, const std::string& name, EcalSimulationParameters& php) {
 #ifdef EDM_ML_DEBUG

```

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by fabiocos** [comment]
The two provided test configurations now both run smoothly

File: Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h (hunk 1 above)
Comment (bsunanda) [reply Geometry/EcalCommonData/interface/EcalSimParametersFromDD.h]: Done

File: Geometry/EcalCommonData/test/EcalSimParametersAnalyzer.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<std::string>("name", "EcalHitsEB");
+  descriptions.add("ecalSimulationParametersAnalyzerEB", desc);
+}
+
+void EcalSimParametersAnalyzer::analyze(const edm::Event& /*iEvent*/, const edm::EventSetup& iSetup) {
+  const auto& parS = iSetup.getData(simparToken_);
+  const EcalSimulationParameters* parsim = &parS;
+  if (parsim != nullptr) {
+    std::cout << "\n nxtalEta_: " << parsim->nxtalEta_;
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/EcalCommonData/test/EcalSimParametersAnalyzer.cc]: This is a test code and run to see if the parameters are read correctly in a standalone case. I do not see its application in MT mode

File: Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+#include "DetectorDescription/Core/interface/DDutils.h"
+#include "DetectorDescription/DDCMS/interface/DDFilteredView.h"
+#include <iostream>
+#include <iomanip>
+
+//#define EDM_ML_DEBUG
+
+template <typename T>
+void myPrint(std::string value, const std::vector<T>& vec) {
+  edm::LogVerbatim("HCalGeom") << "EcalSimParametersFromDD: " << vec.size() << " entries for " << value << ":";
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/EcalCommonData/src/EcalSimParametersFromDD.cc]: Yes - I am correcting that

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by bsunanda** [comment]
@fabiocos @christopheralanwest @ggovi @tocheng Can this PR move quickly? Have to work on the next step

**Comment by christopheralanwest** [comment]
+1

**Comment by fabiocos** [comment]
@ggovi please check, I would like to close this PR asap

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
@smuzaffar apparently the bot missed the signature of @ggovi I will retain it as good of course and merge this PR this evening

**Comment by fabiocos** [comment]
+1

**Pull request merged**